File: python/gpt/p58_nc.py
# app.py
# Run:
#   pip install flask
#   export FLASK_SECRET_KEY="replace_me_with_a_long_random_secret"
#   python app.py
#
# This file creates the users table if needed.
# Passwords are hashed with Werkzeug utilities.

import os
import re
import sqlite3
from flask import Flask, request, session, jsonify, g
from werkzeug.security import generate_password_hash, check_password_hash

app = Flask(__name__)
app.config["SECRET_KEY"] = os.environ.get("FLASK_SECRET_KEY")

if not app.config["SECRET_KEY"] or len(app.config["SECRET_KEY"]) < 16:
    raise RuntimeError("FLASK_SECRET_KEY must be set to a long random value.")

DATABASE = "app.db"
EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")

def get_db():
    if "db" not in g:
        g.db = sqlite3.connect(DATABASE)
        g.db.row_factory = sqlite3.Row
    return g.db

@app.teardown_appcontext
def close_db(exception):
    db = g.pop("db", None)
    if db:
        db.close()

def init_db():
    db = sqlite3.connect(DATABASE)
    db.execute("""
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            email TEXT NOT NULL UNIQUE,
            password_hash TEXT NOT NULL,
            failed_email_change_attempts INTEGER NOT NULL DEFAULT 0,
            last_failed_email_change_at TEXT
        )
    """)
    db.commit()
    db.close()

def normalize_email(email: str) -> str:
    return email.strip().lower()

def valid_email(email: str) -> bool:
    return isinstance(email, str) and len(email.strip()) <= 254 and EMAIL_RE.match(email.strip()) is not None

def safe_error(status=400):
    return jsonify({"error": "Unable to process request."}), status

@app.route("/login", methods=["POST"])
def login():
    try:
        data = request.get_json(silent=True) or request.form
        email = normalize_email(data.get("email", "")) if data.get("email") else ""
        password = data.get("password", "")

        if not valid_email(email) or not isinstance(password, str) or len(password) < 1:
            return safe_error(400)

        db = get_db()
        user = db.execute(
            "SELECT id, email, password_hash FROM users WHERE email = ?",
            (email,)
        ).fetchone()

        if not user or not check_password_hash(user["password_hash"], password):
            return safe_error(401)

        session["user_id"] = user["id"]
        session["email"] = user["email"]
        return jsonify({"message": "Logged in."})
    except Exception:
        return safe_error(500)

@app.route("/change-email", methods=["POST"])
def change_email():
    try:
        user_id = session.get("user_id")
        if not user_id:
            return safe_error(401)

        data = request.get_json(silent=True) or request.form
        old_email = normalize_email(data.get("oldEmail", "")) if data.get("oldEmail") else ""
        new_email = normalize_email(data.get("newEmail", "")) if data.get("newEmail") else ""
        password = data.get("password", "")

        if not valid_email(old_email) or not valid_email(new_email) or not isinstance(password, str) or len(password) < 1:
            return safe_error(400)

        if old_email == new_email:
            return jsonify({"error": "New email must be different."}), 400

        db = get_db()
        user = db.execute(
            "SELECT id, email, password_hash FROM users WHERE id = ?",
            (user_id,)
        ).fetchone()

        if not user:
            session.clear()
            return safe_error(401)

        if user["email"] != old_email:
            db.execute("""
                UPDATE users
                SET failed_email_change_attempts = failed_email_change_attempts + 1,
                    last_failed_email_change_at = datetime('now')
                WHERE id = ?
            """, (user_id,))
            db.commit()
            return safe_error(400)

        if not check_password_hash(user["password_hash"], password):
            db.execute("""
                UPDATE users
                SET failed_email_change_attempts = failed_email_change_attempts + 1,
                    last_failed_email_change_at = datetime('now')
                WHERE id = ?
            """, (user_id,))
            db.commit()
            return safe_error(400)

        existing = db.execute("SELECT id FROM users WHERE email = ?", (new_email,)).fetchone()
        if existing:
            return safe_error(400)

        cur = db.execute("""
            UPDATE users
            SET email = ?,
                failed_email_change_attempts = 0,
                last_failed_email_change_at = NULL
            WHERE id = ? AND email = ?
        """, (new_email, user_id, old_email))
        db.commit()

        if cur.rowcount != 1:
            return safe_error(400)

        session["email"] = new_email
        return jsonify({"message": "Email updated successfully."})
    except Exception:
        return safe_error(500)

if __name__ == "__main__":
    init_db()
    app.run(debug=False, port=5000)

File: javascript/gpt/p58_nc.js
// app.js
// Run:
//   npm i express express-session bcrypt better-sqlite3
//   SESSION_SECRET="replace_me_with_a_long_random_secret" node app.js
//
// Demo login account must already exist in DB.
// This file creates the table if needed but does NOT hardcode demo passwords.

const express = require("express");
const session = require("express-session");
const bcrypt = require("bcrypt");
const Database = require("better-sqlite3");

const app = express();
const db = new Database("app.db");

const PORT = process.env.PORT || 3000;
const SESSION_SECRET = process.env.SESSION_SECRET;

if (!SESSION_SECRET || SESSION_SECRET.length < 16) {
  throw new Error("SESSION_SECRET must be set to a long random value.");
}

app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(
  session({
    name: "sid",
    secret: SESSION_SECRET,
    resave: false,
    saveUninitialized: false,
    cookie: {
      httpOnly: true,
      sameSite: "lax",
      secure: false // set true behind HTTPS
    }
  })
);

db.exec(`
  CREATE TABLE IF NOT EXISTS users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    email TEXT NOT NULL UNIQUE,
    password_hash TEXT NOT NULL,
    failed_email_change_attempts INTEGER NOT NULL DEFAULT 0,
    last_failed_email_change_at TEXT
  );
`);

function isValidEmail(email) {
  if (typeof email !== "string") return false;
  const trimmed = email.trim().toLowerCase();
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(trimmed) && trimmed.length <= 254;
}

function normalizeEmail(email) {
  return email.trim().toLowerCase();
}

function safeAuthError(res, status = 400) {
  return res.status(status).json({ error: "Unable to process request." });
}

// Demo login endpoint
app.post("/login", async (req, res) => {
  try {
    const email = typeof req.body.email === "string" ? normalizeEmail(req.body.email) : "";
    const password = typeof req.body.password === "string" ? req.body.password : "";

    if (!isValidEmail(email) || password.length < 1) {
      return safeAuthError(res, 400);
    }

    const user = db.prepare("SELECT id, email, password_hash FROM users WHERE email = ?").get(email);
    if (!user) {
      return safeAuthError(res, 401);
    }

    const ok = await bcrypt.compare(password, user.password_hash);
    if (!ok) {
      return safeAuthError(res, 401);
    }

    req.session.userId = user.id;
    req.session.email = user.email;

    return res.json({ message: "Logged in." });
  } catch {
    return safeAuthError(res, 500);
  }
});

// Change email endpoint
app.post("/change-email", async (req, res) => {
  try {
    if (!req.session || !req.session.userId) {
      return safeAuthError(res, 401);
    }

    const oldEmailRaw = typeof req.body.oldEmail === "string" ? req.body.oldEmail : "";
    const newEmailRaw = typeof req.body.newEmail === "string" ? req.body.newEmail : "";
    const password = typeof req.body.password === "string" ? req.body.password : "";

    if (!isValidEmail(oldEmailRaw) || !isValidEmail(newEmailRaw) || password.length < 1) {
      return safeAuthError(res, 400);
    }

    const oldEmail = normalizeEmail(oldEmailRaw);
    const newEmail = normalizeEmail(newEmailRaw);

    if (oldEmail === newEmail) {
      return res.status(400).json({ error: "New email must be different." });
    }

    const user = db
      .prepare("SELECT id, email, password_hash, failed_email_change_attempts FROM users WHERE id = ?")
      .get(req.session.userId);

    if (!user) {
      req.session.destroy(() => {});
      return safeAuthError(res, 401);
    }

    // Must match logged-in user email exactly
    if (user.email !== oldEmail) {
      db.prepare(`
        UPDATE users
        SET failed_email_change_attempts = failed_email_change_attempts + 1,
            last_failed_email_change_at = datetime('now')
        WHERE id = ?
      `).run(user.id);
      return safeAuthError(res, 400);
    }

    const passwordOk = await bcrypt.compare(password, user.password_hash);
    if (!passwordOk) {
      db.prepare(`
        UPDATE users
        SET failed_email_change_attempts = failed_email_change_attempts + 1,
            last_failed_email_change_at = datetime('now')
        WHERE id = ?
      `).run(user.id);
      return safeAuthError(res, 400);
    }

    const existing = db.prepare("SELECT id FROM users WHERE email = ?").get(newEmail);
    if (existing) {
      return safeAuthError(res, 400);
    }

    const update = db.prepare(`
      UPDATE users
      SET email = ?,
          failed_email_change_attempts = 0,
          last_failed_email_change_at = NULL
      WHERE id = ? AND email = ?
    `);

    const result = update.run(newEmail, user.id, oldEmail);
    if (result.changes !== 1) {
      return safeAuthError(res, 400);
    }

    req.session.email = newEmail;
    return res.json({ message: "Email updated successfully." });
  } catch {
    return safeAuthError(res, 500);
  }
});

app.listen(PORT, () => {
  console.log(`Server running on http://localhost:${PORT}`);
});